crates/e2e/src/setup/onchain_components.rs

```diff
@@ -66,6 +68,55 @@ macro_rules! tx_safe {
     }};
 }
 
+pub struct GnosisSafeInfrastructure {
+    pub factory: GnosisSafeProxyFactory,
+    pub fallback: GnosisSafeCompatibilityFallbackHandler,
+    pub singleton: GnosisSafe,
+    web3: Web3,
+}
+
+impl GnosisSafeInfrastructure {
+    pub async fn new(web3: &Web3) -> Self {
+        let singleton = GnosisSafe::builder(web3).deploy().await.unwrap();
```

Comment:
Why is this called singleton?
Is it because it gets created with `create2` and always ends up with the same address? What happens if it gets deployed multiple times?

Reply:
> Why is this called singleton?

It is worth mentioning that this is the terminology used by the Safe contracts/docs/etc., so it makes sense to keep it in our code as well.